virttest/qemu_devices/qdevices.py

```diff
@@ -2352,6 +2356,127 @@ def __eq__(self, other):
         return False
 
 
+class QPasstDev(QDaemonDev):
+    """
+    Passt pseudo device.
+    """
+
+    def __init__(self, aobject, binary, sock_path, rules, mtu=None):
+        """
+        :param aobject: The aobject of virtiofs daemon.
+        :type aobject: str
+        :param binary: The binary of virtiofs daemon.
+        :type binary: str
+        :param sock_path: The socket path of of virtiofs daemon.
+        :type sock_path: str
+        :param source: The source of virtiofs daemon.
+        :type source: str
+        :param extra_options: The external options of virtiofs daemon.
+        :type extra_options: str
+        :param enable_debug_mode: Enable debug mode of virtiofs daemon.
+        :type enable_debug_mode: bool
+        """
+        super().__init__(
+            "passt",
+            aobject=aobject,
+            child_bus=QUnixSocketBus(sock_path, aobject),
+        )
+        self.set_param("binary", binary)
+        self.set_param("sock_path", sock_path)
+        self.set_param("rules", rules)
+        self.set_param("mtu", mtu)
+
+        uuid = uuid4()
+        # let's place the pid file within the same dir
+        self._pid_file = os.path.join(os.path.dirname(sock_path), f"passt-{uuid}.pid")
+        log_filename = f"passt-{uuid}.log"
+        # deal with the permission issue against selinux
+        self._tmp_log_file = os.path.join(
+            "/var/run/user", f"{os.getuid()}", log_filename
+        )
+        self._log_file = utils_logfile.get_log_filename(log_filename)
+
+    def _get_pid(self):
+        try:
+            with open(self._pid_file) as pid_f:
+                return int(pid_f.read().strip())
+        except FileNotFoundError:
+            return None
+
+    def _remove_pid_file(self):
+        if not os.path.exists(self._pid_file):
+            return
+        try:
+            os.remove(self._pid_file)
+        except FileNotFoundError:
+            pass
+
+    def _save_log_file(self):
+        if not os.path.exists(self._tmp_log_file):
+            return
+        try:
+            shutil.move(self._tmp_log_file, self._log_file)
+        except:
+            pass
+
+    def is_daemon_alive(self):
+        pid = self._get_pid()
+        if pid is None:
+            return False
+        return process.pid_exists(pid)
+
+    def _render_rules(self):
+        cmd = ""
+        for rule in self.get_param("rules", "").split():
+            # skip empty
+            if not rule:
+                continue
+            proto, _, spec = rule.rpartition("@")
+            if proto.lower() in ("tcp", "t", ""):
+                cmd += f" --tcp-ports {spec}"
+            elif proto.lower() in ("udp", "u"):
+                cmd += f" --udp-ports {spec}"
+            else:
+                raise ValueError(f"unsupported protocol {proto}")
+        return cmd
+
+    def start_daemon(self):
+        """Start the passt daemon."""
+        cmd = "%s --one-off --socket %s" % (
```

Comment:
I would recommend adding `--foreground` by default, do not let `QDaemonDev` manage a background process, I think it's outside the scope of its design. And with this, `process.pid_exists(pid)` may not a must, can just use the original one.

But let's listen to other peoples' voices.

Reply:
Put https://github.com/avocado-framework/avocado-vt/pull/2929 as a reference. In qdevices.py, we already use aexpect.run_bg, any process will run in backaground, not need to background again.

@yanglei-rh, feel free leave your comment.

Reply:
And a unit test like this:
```python
>>> passt1 = client.run_bg("passt")
>>> passt2 = client.run_bg("passt -f")
>>> passt1.is_alive()
False
>>> passt2.is_alive()
True
```

Reply:
Hi @PaulYuuu Sorry for the late reply, I agree with your point.

Reply:
Hi @PaulYuuu Thanks for pointing it out
Hi @luckyh Maybe we can directly to use `aexpect.run_bg` instead of call `QDaemonDev`?

Reply:
As daemonized processes are always configured with extra settings [1], I'd suggest keeping the daemon mode in favor of problem detection, in case the user made SELinux enforced.

What do you think @PaulYuuu ?

[1] https://passt.top/passt/tree/util.c#n449